pkg/controller/core/clusterqueue_controller.go

```diff
@@ -497,3 +552,93 @@ func (r *ClusterQueueReconciler) updateCqStatusIfChanged(
 	}
 	return nil
 }
+
+func (r *ClusterQueueReconciler) getWorkloadsStatus(cq *kueue.ClusterQueue) *kueue.ClusterQueuePendingWorkloadsStatus {
+	if cq.Status.PendingWorkloadsStatus == nil {
+		return &kueue.ClusterQueuePendingWorkloadsStatus{
+			Head:           r.qManager.GetSnapshot(cq.Name),
+			LastChangeTime: metav1.Time{Time: time.Now()},
+		}
+	}
+	if time.Since(cq.Status.PendingWorkloadsStatus.LastChangeTime.Time) >= r.queueVisibilityUpdateInterval {
+		pendingWorkloads := r.qManager.GetSnapshot(cq.Name)
+		if !equality.Semantic.DeepEqual(cq.Status.PendingWorkloadsStatus.Head, pendingWorkloads) {
+			return &kueue.ClusterQueuePendingWorkloadsStatus{
+				Head:           pendingWorkloads,
+				LastChangeTime: metav1.Time{Time: time.Now()},
+			}
+		}
+	}
+	return cq.Status.PendingWorkloadsStatus
+}
+
+func (r *ClusterQueueReconciler) Start(ctx context.Context) error {
+	defer r.snapshotsQueue.ShutDown()
+
+	for i := 0; i < snapshotWorkers; i++ {
+		go wait.UntilWithContext(ctx, r.takeSnapshot, r.queueVisibilityUpdateInterval)
+	}
+
+	go wait.UntilWithContext(ctx, r.enqueueTakeSnapshot, r.queueVisibilityUpdateInterval)
+
+	<-ctx.Done()
+
+	return nil
+}
+
+func (r *ClusterQueueReconciler) enqueueTakeSnapshot(ctx context.Context) {
+	for cq := range r.qManager.GetClusterQueues() {
+		r.snapshotsQueue.Add(cq)
+	}
+}
+
+func (r *ClusterQueueReconciler) takeSnapshot(ctx context.Context) {
+	for r.processNextSnapshot(ctx) {
+	}
+}
+
+func (r *ClusterQueueReconciler) processNextSnapshot(ctx context.Context) bool {
+	log := ctrl.LoggerFrom(ctx).WithName("processNextSnapshot")
+
+	key, quit := r.snapshotsQueue.Get()
+	if quit {
+		return false
+	}
+
+	startTime := time.Now()
+	defer func() {
+		log.V(2).Info("Finished snapshot job", "key", key, "elapsed", time.Since(startTime))
+	}()
+
+	defer r.snapshotsQueue.Done(key)
+
+	workloads := make([]kueue.ClusterQueuePendingWorkload, 0)
+	if elements, ok := r.getSnapshotFromClusterQueue(key); ok {
+		for index, info := range elements {
+			if int32(index) >= r.queueVisibilityClusterQueuesMaxCount {
+				break
+			}
+			if info == nil {
+				continue
+			}
+			workloads = append(workloads, kueue.ClusterQueuePendingWorkload{
+				Name:      info.Obj.Name,
+				Namespace: info.Obj.Namespace,
+			})
+		}
+	}
+	r.qManager.SetSnapshot(key.(string), workloads)
+
+	return true
+}
+
+func (r *ClusterQueueReconciler) getSnapshotFromClusterQueue(key interface{}) ([]*workload.Info, bool) {
+	clusterQueues := r.qManager.GetClusterQueues()
+	if len(clusterQueues) == 0 {
```

Comment:
Here we might be checking the state of CQ map, but we no longer have the lock. We should call  `r.qManager.GetClusterQueue(cqName)`. Then, if there is the queue, then we call `cq.Snapshot()` which will take the cq-level lock.